packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp

```diff
@@ -659,17 +662,11 @@ void generalWriteString(const std::string& str, std::ostream& yaml)
 
 void generalWriteDouble(double d, std::ostream& yaml)
 {
-  yaml << std::showpoint << std::setprecision(8);
-  if(d < 1e6 && d > 1e-5)
-  {
-    //use regular notation
-    yaml << d;
-  }
-  else
-  {
-    yaml << std::scientific << d;
-  }
-  yaml << std::fixed;
+  std::ios saved_state(0);
+  saved_state.copyfmt(yaml);
+  yaml << std::noshowpoint << std::scientific << std::setprecision(15);
```

Comment:
Actually, I would go further and say that the only way I would feel comfortable guaranteeing binary restart is with hexadecimal encodings, e.g. C++11's `std::hexfloat`. Unfortunately, that format is not so good for human readability. This PR is a balance between accuracy and readability. The goal is to keep errors low enough that regression tests don't start failing due to XML->YAML conversion, but I really don't expect binary exactness.